### PR TITLE
Java.Lang.String

### PR DESCRIPTION
**Pull request by TuanAnhNguyen69**

### CODE REVIEW

File: java/lang/String/String.cpp [hunk 1]
```diff
@@ -68,104 +71,113 @@ String::String(Array<byte> &bytes) {
...
 }
 
-/**
- * String character at index
- *
- * @param index
- * @return String
- */
-char String::charAt(int index) const {
+char String::charAt(int index) const{
```
Comment (ghost) [inline java/lang/String/String.cpp]: Missing space before {

File: java/lang/String/String.cpp [hunk 2]
```diff
@@ -175,43 +187,25 @@ Array<byte> String::getBytes() const {
...
 
-/**
- * String endswith a suffix
- * @param suffix
- * @return
- */
-boolean String::endsWith(const String &suffix) const {
-	return ( string_endswith(this->original, suffix.original));
+boolean String::endsWith(const String &suffixString) const {
+	return static_cast<boolean>(string_endswith(this->original, suffixString.original));
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 3]
```diff
@@ -220,185 +214,138 @@ String String::fromCharArray(Array<char> &chars) {
...
 	string currentReversedString = string_reverse(this->toString());
 	int result = string_index(currentReversedString, reversedString, 1);
 	free(reversedString);
 	free(currentReversedString);
 	if (result == NOT_FOUND) {
 		return result;
 	}
-	//Re-calculate first character of str
-	result = this->size - ( result + str.size );
+	//Re-calculate first character of subString
```
Comment (ghost) [inline java/lang/String/String.cpp]: Should have a space between // and comment

File: java/lang/String/String.cpp [hunk 4]
```diff
@@ -220,185 +214,138 @@ String String::fromCharArray(Array<char> &chars) {
...
+    if (fromIndex < 0) {
+        return -1;
+    }
+    if (fromIndex > this->size - 1) {
+        return this->lastIndexOf(subString);
+    }
+    string thisStringReversed = string_reverse(this->original);
+	string subStringFromIndex = &(thisStringReversed)[ this->size - fromIndex - subString.size]; // get subString start fromIndex
+	string reversedString = string_reverse(subString.toString());
+	//string currentReversedString = string_reverse(subStringFromIndex);
```
Comment (ghost) [inline java/lang/String/String.cpp]: Should have a space between // and comment

File: java/lang/String/String.cpp [hunk 5]
```diff
@@ -410,25 +357,13 @@ String String::replace(char oldChar, char newChar) const {
...
- * String replace all matches by replacement
- *
- * @param regex
- * @param replacement
- * @return String
- */
 String String::replaceAll(String regex, String replacement) const {
-	// TODO fix this later
-	return "";
+	// TODO fix this later, temporary use replace, need Pattern
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.cpp [hunk 6]
```diff
@@ -410,25 +357,13 @@ String String::replace(char oldChar, char newChar) const {
...
 }
 
-/**
- * Split string by regex
- *
- * @param regex
- * @return Array<String>
- */
 Array<String> String::split(String regex) const {
+    // TODO fix this later, temporary use replace, need Pattern
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.cpp [hunk 7]
```diff
@@ -446,31 +381,18 @@ Array<String> String::split(String regex) const {
...
 
-/**
- * String starts with a prefix
- *
- * @param prefix
- * @return boolean
- */
 boolean String::startsWith(String prefix) const {
-	return string_startswith(this->original, prefix.original);
+	return static_cast<boolean>(string_startswith(this->original, prefix.original));
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 8]
```diff
@@ -446,31 +381,18 @@ Array<String> String::split(String regex) const {
...
- *
- * @param prefix
- * @param from t
- * @return boolean
- */
-boolean String::startsWith(String prefix, int toffset) const {
-	if (this->original == NULL || prefix.original == NULL || toffset < 0) {
-		return FALSE;
+boolean String::startsWith(String prefix, int thisOffset) const {
+	if (this->original == nullptr || prefix.original == nullptr || thisOffset < 0) {
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 9]
```diff
@@ -446,31 +381,18 @@ Array<String> String::split(String regex) const {
...
+		return false;
 	}
 	int originalLength = length_pointer_char(this->original);
 	int prefixLength = length_pointer_char(prefix.original);
-	if (originalLength < prefixLength || toffset > ( originalLength - prefixLength )) {
-		return FALSE;
+	if (originalLength < prefixLength || thisOffset > (originalLength - prefixLength)) {
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 10]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
- * @return boolean
- */
 boolean String::operator>=(const String &target) const {
-	if (strcmp(this->original, target.toString()) < 0) {
-		return false;
+    return strcmp(this->original, target.toString()) >= 0;
+}
+
+boolean String::contentEquals(const CharSequence &charSequence) {
+    // TODO instanceof return false
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.cpp [hunk 11]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+    if (hashCode == 0 && this->size > 0) {
+        for (int i = 0; i < this->size; i++) {
+            hashCode = 31 * hashCode + this->original[i];
+        }
+        this->hash = hashCode;
+    }
+    return hashCode;
+}
+
+boolean String::regionMatches(int thisOffset, String otherString, int otherOffset, int len) {
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 12]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+	return this->regionMatches(false, thisOffset, otherString, otherOffset, len);
+}
+
+boolean String::regionMatches(boolean ignoreCase, int thisOffset, String otherString, int otherOffset, int len) {
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 13]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+	if (ignoreCase) {
+		return thisString.compareToIgnoreCase(otherString) == 0;
 	}
-	return true;
-}
\ No newline at end of file
+	return thisString.compareTo(otherString) == 0;
+}
+
+void String::getChars(int sourceBegin, int sourceEnd, Array<char> &destination, int destinationBegin) {
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 14]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+    string pointerHolder = string_replace(this->original, oldString, newString);
+    String result = pointerHolder;
+    free(pointerHolder);
+    free(oldString);
+    free(newString);
+    return result;
+}
+
+String String::replaceFirst(String regex, String replacement) const {
+    // TODO fix this later, temporary, need Pattern
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.cpp [hunk 15]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+    String stringWithFirstRegex = this->subString(0, this->indexOf(regex) + regex.length());
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 16]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+    String remainString = this->getStringFromIndex(this->indexOf(regex) + regex.length());
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 17]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+    stringWithFirstRegex = stringWithFirstRegex.replace(regex, replacement);
+    return stringWithFirstRegex + remainString;
+}
+
+Array<String> String::split(String regex, int limit) const {
+    // TODO fix this later, temporary, need Pattern
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.cpp [hunk 18]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+    Array<String> stringArrayNoLimit = this->split(regex);
+    if (limit == 1) {
+        return Array<String>{*this};
+    }
+    if (limit > stringArrayNoLimit.length || limit <= 0) {
+        return stringArrayNoLimit;
+    }
+    int indexOfRegexBelowLimit = string_index(this->original, regex.toString(), limit - 1);
```
Comment (ghost) [inline java/lang/String/String.cpp:703]: Please split or shorten this line.

File: java/lang/String/String.cpp [hunk 19]
```diff
@@ -508,341 +426,283 @@ Array<char> String::toCharArray() const {
...
+    String remainString = this->getStringFromIndex(indexOfRegexBelowLimit + regex.length());
```
Comment (ghost) [inline java/lang/String/String.cpp]: Please split or shorten this line.

File: java/lang/String/String.hpp [hunk 20]
```diff
@@ -35,121 +39,949 @@
...
+            String(const StringBuffer &stringBuffer);
+
+            /**
+             * Constructs a new String by decoding the specified array of bytes
+             * using the specified charset.
+             *
+             * @param byteArray
+             * @param charsetName
+             */
+            // TODO need CharSet
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 21]
```diff
@@ -35,121 +39,949 @@
...
+            // String(Array<byte> byteArray, String charsetName);
+
+            /**
+             * Constructs a new String by decoding the specified array of bytes
+             * using the specified charset.
+             *
+             * @param byteArray
+             * @param charset
+             */
+            // TODO need CharSet
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 22]
```diff
@@ -35,121 +39,949 @@
...
+             *
+             * @param codePoints
+             * @param offset
+             * @param count
+             * @throw IllegalArgumentException If any invalid Unicode
+             * code point is found in codePoints
+             * @throw IndexOutOfBoundsException If the offset and count
+             * arguments index characters outside the bounds of the codePoints array
+             */
+            // TODO need CharSet
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 23]
```diff
@@ -35,121 +39,949 @@
...
+            //String(Array<int> codePoints, int offset, int count);
```
Comment (ghost) [inline java/lang/String/String.hpp]: Should have a space between // and comment

File: java/lang/String/String.hpp [hunk 24]
```diff
@@ -35,121 +39,949 @@
...
+             * subarray of bytes using the specified charset
+             *
+             * @param byteArray
+             * @param offset
+             * @param length
+             * @param charset
+             * @throw IndexOutOfBoundsException If the offset and length arguments index
+             * characters outside the bounds of the bytes array
+             */
+            // TODO need CharSet
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 25]
```diff
@@ -35,121 +39,949 @@
...
+             *
+             * @param bytes
+             * @param offset
+             * @param length
+             * @param charsetName
+             * @throw java.io.UnsupportedEncodingException If the named charset is not supported
+             * @throw IndexOutOfBoundsException If the offset and length arguments index
+             * characters outside the bounds of the bytes array
+             */
+            // TODO need CharSet
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 26]
```diff
@@ -35,121 +39,949 @@
...
+            //String(Array<byte> &byteArray, int offset, int length, String charsetName);
```
Comment (ghost) [inline java/lang/String/String.hpp]: Should have a space between // and comment

File: java/lang/String/String.hpp [hunk 27]
```diff
@@ -35,121 +39,949 @@
...
+			~String();
+
+        private:
+            /**
+             * Bounds check the byte array and requested offset & length values
+             * @param byteArray
+             * @param offset
+             * @param length
+             */
+            //static void checkBounds(Array<byte> &byteArray, int offset, int length);
```
Comment (ghost) [inline java/lang/String/String.hpp]: Should have a space between // and comment

File: java/lang/String/String.hpp [hunk 28]
```diff
@@ -35,121 +39,949 @@
...
+             */
+			char charAt(int index) const override;
+
+            /**
+             * Returns the character (Unicode code point) at the specified index.
+             *
+             * @param index
+             * @return int
+             */
+            // TODO need Character
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 29]
```diff
@@ -35,121 +39,949 @@
...
+			// int codePointAt(int index);
+
+            /**
+             * Returns the character (Unicode code point) before the specified index.
+             *
+             * @param index
+             * @return int
+             */
+            // TODO need Character
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 30]
```diff
@@ -35,121 +39,949 @@
...
+
+            /**
+             * Returns the number of Unicode code points
+             * in the specified text range of this String.
+             *
+             * @param beginIndex
+             * @param endIndex
+             * @return int
+             */
+            // TODO need Character
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 31]
```diff
@@ -35,121 +39,949 @@
...
+			Array<byte> getBytes() const;
+
+            /**
+             * Encodes this String into a sequence of bytes using the named charset,
+             * storing the result into a new byte array.
+             *
+             * @param charSetName
+             * @return
+             */
+            //Array<byte> getBytes(String charSetName) const;
```
Comment (ghost) [inline java/lang/String/String.hpp]: Should have a space between // and comment

File: java/lang/String/String.hpp [hunk 32]
```diff
@@ -35,121 +39,949 @@
...
+             * @param destination
+             * @param destinationBegin
+             * @throw StringIndexOutOfBoundsException If :
+             * sourceBegin is negative.
+             * sourceBegin is greater than srcEnd
+             * sourceEnd is greater than the length of this string
+             * destinationBegin is negative
+             * destinationBegin + (srcEnd - srcBegin) is larger than destination.length
+             */
+            void getChars(int sourceBegin, int sourceEnd, Array<char> &destination, int destinationBegin);
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please split or shorten this line.

File: java/lang/String/String.hpp [hunk 33]
```diff
@@ -35,121 +39,949 @@
...
+			int length() const override;
+
+            /**
+             * Tells whether or not this string matches the given regular expression.
+             *
+             * @param regex
+             * @throw java.util.regex.PatternSyntaxException if the regular expression's syntax is invalid
+             * @return true if, and only if, this string matches the given regular expression
+             */
+            // TODO need Pattern
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 34]
```diff
@@ -35,121 +39,949 @@
...
+             * @param index
+             * @param codePointOffset
+             * @throw IndexOutOfBoundsException if index is negative or larger then the length of this String,
+             * or if codePointOffset is positive and the substring starting with index
+             * has fewer than codePointOffset code points,
+             * or if codePointOffset is negative and the substring before index has fewer than
+             * the absolute value of codePointOffset code points.
+             * @return the index within this String
+             */
+            // TODO need Character
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/String/String.hpp [hunk 35]
```diff
@@ -35,121 +39,949 @@
...
+             * Tests if two string regions are equal.
+             *
+             * @param thisOffset
+             * @param otherString
+             * @param otherOffset
+             * @param len
+             * @return true if the specified subregion of this String exactly
+             * matches the specified subregion of the String argument; false otherwise.
+             */
+            boolean regionMatches(int thisOffset, String otherString, int otherOffset, int len);
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please split or shorten this line.

File: java/lang/String/String.hpp [hunk 36]
```diff
@@ -35,121 +39,949 @@
...
+             * @param ignoreCase
+             * @param thisOffset
+             * @param otherString
+             * @param otherOffset
+             * @param len
+             * @return true if the specified subregion of this String matches the specified
+             * subregion of the String argument; false otherwise.
+             * Whether the matching is exact or case insensitive depends on the ignoreCase argument.
+             */
+            boolean regionMatches(boolean ignoreCase, int thisOffset, String otherString, int otherOffset, int len);
```
Comment (ghost) [inline java/lang/String/String.hpp]: Please split or shorten this line.

File: java/lang/String/String.hpp [hunk 37]
```diff
@@ -35,121 +39,949 @@
...
+            /**
+             * Returns a character sequence that is a subsequence of this sequence.
+             *
+             * @param beginIndex
+             * @param endIndex
+             * @throw IndexOutOfBoundsException if beginIndex or endIndex is negative,
+             * if endIndex is greater than length(), or if beginIndex is greater than endIndex
+             * @return the specified subsequence.
+             */
+            //CharSequence subSequence(int beginIndex, int endIndex);
```
Comment (ghost) [inline java/lang/String/String.hpp]: Should have a space between // and comment

File: java/lang/String/String.hpp [hunk 38]
```diff
@@ -35,121 +39,949 @@
...
+			 * @return a String contain value of target1 and target2
+			 */
+			friend String operator+(const_string target1, String const &target2) {
+				String result;
+				result = target1;
+				result += target2;
+				return result;
+			}
+		};
+	}
```
Comment (ghost) [inline java/lang/String/String.hpp]: Namespace should be terminated with "// namespace Lang"

File: java/lang/String/StringTest.cpp [hunk 39]
```diff
@@ -112,8 +208,7 @@ TEST (JavaLang, StringConcat) {
 	ASSERT_STR("Food Tiny Hello World", ( textConcat0 + textConcat1 + textConcat2 ).toString());
```
Comment (ghost) [inline java/lang/String/StringTest.cpp]: Extra space after (. Extra space before )

File: java/lang/String/StringTest.cpp [hunk 40]
```diff
@@ -180,24 +362,84 @@ TEST (JavaLang, StringIndexOf) {
...
+
+    ASSERT_EQUAL(0, validString2.lastIndexOf(subString2, 1));
+
+    ASSERT_EQUAL(19, validString2.lastIndexOf(subString2, 100));
+
+    ASSERT_EQUAL(-1, validString2.lastIndexOf(subString2, -1));
+
+	ASSERT_EQUAL(19, validString2.lastIndexOf(subString2, 20));
+}
+// TODO getChar run right but need Arrays.toString() to test
```
Comment (ghost) [inline java/lang/String/StringTest.cpp]: Please add your name to TODO comment like // TODO(yourname):

File: java/lang/StringBuffer/StringBuffer.hpp [hunk 41]
```diff
@@ -485,7 +485,7 @@ namespace Java {
              * @throw IndexOutOfBoundsException - if the offset is invalid.
              * @return a reference to this StringBufferUnSafe
              */
-            StringBufferUnSafe &insert(int destinationOffset, CharSequence *sequence, int start, int end);
+            StringBufferUnSafe &insert(int destinationOffset, CharSequence &sequence, int start, int end);
```
Comment (ghost) [inline java/lang/StringBuffer/StringBuffer.hpp:490]: Please split or shorten this line.

File: java/lang/StringBuffer/StringBuffer.hpp [hunk 42]
```diff
@@ -1089,7 +1089,7 @@ namespace Java {
              * @throw IndexOutOfBoundsException - if the offset is invalid.
              * @return a reference to this StringBuffer
              */
-            StringBuffer &insert(int destinationOffset, CharSequence *sequence, int start, int end);
+            StringBuffer &insert(int destinationOffset, CharSequence &sequence, int start, int end);
```
Comment (ghost) [inline java/lang/StringBuffer/StringBuffer.hpp:1094]: Please split or shorten this line.

File: java/lang/StringBuffer/StringBufferTest.cpp [hunk 43]
```diff
@@ -39,62 +39,62 @@ extern "C" {
 using namespace Java::Lang;
 
 TEST (JavaLang, StringBufferConstructor) {
```
Comment (ghost) [inline java/lang/StringBuffer/StringBufferTest.cpp]: Extra space before ( in function call

File: java/lang/StringBuffer/StringBufferTest.cpp [hunk 44]
```diff
@@ -171,148 +171,146 @@ TEST (JavaLang, StringBufferAppendSubCharArray) {
 }
 
 TEST (JavaLang, StringBufferAppend) {
```
Comment (ghost) [inline java/lang/StringBuffer/StringBufferTest.cpp]: Extra space before ( in function call

File: java/lang/StringBuffer/StringBufferTest.cpp [hunk 45]
```diff
@@ -171,148 +171,146 @@ TEST (JavaLang, StringBufferAppendSubCharArray) {
...
+    // Test null CharSequence
+    StringBuffer nullCharSequenceAppendStringBuffer = StringBuffer("CharSequence is : ");
+
+    CharSequence *nullCharSequenceToAppend = nullptr;
+    nullCharSequenceAppendStringBuffer.append(*nullCharSequenceToAppend);
+    string expectNullCharSequenceAppend = (string)("CharSequence is : null");
+    ASSERT_STR(expectNullCharSequenceAppend, nullCharSequenceAppendStringBuffer.getValue());
 }
 
 TEST (JavaLang, StringBufferInsertSubCharArray) {
```
Comment (ghost) [inline java/lang/StringBuffer/StringBufferTest.cpp]: Extra space before ( in function call

File: java/lang/StringBuffer/StringBufferTest.cpp [hunk 46]
```diff
@@ -367,172 +365,172 @@ TEST (JavaLang, StringBufferInsertSubCharArray) {
 }
 
 TEST (JavaLang, StringBufferInsert) {
```
Comment (ghost) [inline java/lang/StringBuffer/StringBufferTest.cpp]: Extra space before ( in function call

File: java/lang/StringBuffer/StringBufferTest.cpp [hunk 47]
```diff
@@ -367,172 +365,172 @@ TEST (JavaLang, StringBufferInsertSubCharArray) {
...
+    // Test null CharSequence
+    StringBuffer nullCharSequenceInsertWithOffsetStringBuffer = StringBuffer("CharSequence is a  pointer");
+
+    nullCharSequenceInsertWithOffsetStringBuffer.insert(18, *nullCharSequenceToInsert, 0, 3);
+    ASSERT_STR(expectNullCharSequenceInsert, nullCharSequenceInsertStringBuffer.getValue());
+    String *charSequenceString = dynamic_cast<String *>(charSequenceToInsert);
+    delete charSequenceString;
 }
 
 TEST (JavaLang, StringBufferEnsureCapacity) {
```
Comment (ghost) [inline java/lang/StringBuffer/StringBufferTest.cpp]: Extra space before ( in function call

**Review by ghost — DISMISSED** [review]

File: java/lang/String/String.hpp [hunk 48]
```diff
@@ -35,121 +39,984 @@
...
+             * @return true if the character sequence represented by the argument
+             * is a suffix of the character sequence represented by this String;
+             * False otherwise
+             */
+			boolean endsWith(const String &suffixString) const;
+
+            /**
+             * Compares this String to the specified object.
+             *
+             * @tparam T
```
Comment (dthongvl) [inline java/lang/String/String.hpp]: @tparam

File: java/lang/StringBuffer/StringBuffer.hpp [hunk 49]
```diff
@@ -31,7 +31,7 @@
 
 namespace Java {
     namespace Lang {
-        class StringBufferUnSafe : public Object{  //: public CharSequence {
+        class StringBufferUnSafe : public Object, public virtual CharSequence {
```
Comment (dthongvl) [inline java/lang/StringBuffer/StringBuffer.hpp]: Please add line break

File: kernel/string.h [hunk 50]
```diff
@@ -59,6 +60,7 @@ int string_equals(char *target1, char *target2);
 char *string_concat_asm(char *target, char *subtarget);
 char *string_reverse(char *target);
 int string_matches(char *target, char *regex);
+char *string_format(const char* format, va_list args);
```
Comment (dthongvl) [inline kernel/string.h]: Please remove this

**Review by dthongvl — CHANGES_REQUESTED** [review]

**Review by dthongvl — APPROVED** [review]

**Review by loint — APPROVED** [review]

**Pull request merged**